### PR TITLE
wip upgrade to 0.7

### PR DESCRIPTION
**Pull request by KristofferC**

### CODE REVIEW

**Comment by KristofferC** [comment]
Tagging some people who might help with testing, @carlomontec, @tribut, @chriscoey, @ianwilliamson. Grateful for any help testing this, the lack of CI makes upgrading very hard.

This should also support v6 of Pardiso.

**Comment by carlomontec** [comment]
@KristofferC Gladly. Should I install as `add Pardiso#kc/upgrade` , in order to test it?

**Comment by KristofferC** [comment]
Yes :)

**Comment by carlomontec** [comment]
Thanks a lot for your time. 

Using the binary Version 0.7.0 (2018-08-08 06:46 UTC) under Ubuntu 16.04 with MKL installed I get the following:

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0 (2018-08-08 06:46 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

julia> versioninfo()
Julia Version 0.7.0
Commit a4cb80f3ed (2018-08-08 06:46 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 4

(v0.7) pkg> add Pardiso#kc/upgrade
  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Updating git-repo `https://github.com/JuliaSparse/Pardiso.jl.git`
 Resolving package versions...
  Updating `~/.julia/environments/v0.7/Project.toml`
 [no changes]
  Updating `~/.julia/environments/v0.7/Manifest.toml`
 [no changes]

julia> using Pardiso
┌ Error: MKL Pardiso did not manage to load, error thrown was: UndefVarError: libmkl_gd not defined
└ @ Pardiso ~/.julia/packages/Pardiso/Gw5aD/src/Pardiso.jl:126

```

Same procedure under V0.6.4 (compiled from source with MKL) seems to be working.

**Comment by KristofferC** [comment]
Indeed, please `pkg> update` and try again.

File: test/runtests.jl [hunk 1]
```diff
@@ -34,8 +36,7 @@ for pardiso_type in psolvers
         X = similar(B)
 
         # Test unsymmetric, herm indef, herm posdef and symmetric
-        for A in SparseMatrixCSC[A1, A1 + A1', A1'A1, A1 + A1.']
-
+        for A in SparseMatrixCSC[A1, A1 + A1', A1'A1, copy(transpose(A1 + A1))]
```
Comment (photicus) [inline test/runtests.jl]: Should this be `copy(A1 + transpose(A1))`  instead of `copy(transpose(A1 + A1))` ?

**Comment by photicus** [comment]
I'm currently seeing all the tests pass on MKL for Linux (17.0.0.098) mac (17.2.163), and windows (17.2.187). Thanks!

File: test/runtests.jl (hunk 1 above)
Comment (KristofferC) [reply test/runtests.jl]: Indeed, thanks.

**Comment by ianwilliamson** [comment]
Thanks for working on this! On Ubuntu 64bit and Pardiso 6 I get-
```
┌ Error: Pardiso did not manage to load, error thrown was: could not load library "libblas"
│ libblas.so: cannot open shared object file: No such file or directory
└ @ Pardiso ~/.julia/packages/Pardiso/Sakgl/src/Pardiso.jl:159
Testing DataType[]
ERROR: LoadError: No Pardiso library managed to load. Unable to run tests.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at /home/---/.julia/packages/Pardiso/Sakgl/test/runtests.jl:18
 [3] include at ./boot.jl:317 [inlined]
 [4] include_relative(::Module, ::String) at ./loading.jl:1038
 [5] include(::Module, ::String) at ./sysimg.jl:29
 [6] include(::String) at ./client.jl:398
 [7] top-level scope at none:0
in expression starting at /home/---/.julia/packages/Pardiso/Sakgl/test/runtests.jl:17
ERROR: Package Pardiso errored during testing
```

BTW, to build are we supposed to just add the binary to `~/.julia/packages/Pardiso/<random_string>/deps/`? What if there are multiple <random_string> folders?

**Comment by KristofferC** [comment]
Yeah, that is a bit awkward. You can find the correct place by doing `import Pardiso; @show pathof(Pardiso)`. But we should probably think of a better installation method.

**Comment by ianwilliamson** [comment]
In principal you could put the .so file anywhere right? Maybe follow what PyCall/PyPlot does and use something like `ENV['PARDISO']`?

**Comment by KristofferC** [comment]
Yeah, that is the correct way to do it now when new package versions are installed at new locations.

**Comment by KristofferC** [comment]
@ianwilliamson Do you have a Blas library installed? 

I used the guide at https://github.com/xianyi/OpenBLAS/wiki/Precompiled-installation-packages in a new Virtual Machine with Ubuntu and it seemed Julia found it after that.

**Comment by carlomontec** [comment]
Now I get all tests passing with the setup aforementioned. Thanks!!

**Comment by KristofferC** [comment]
Ok, so MKL seems solid. Thanks!

**Comment by ianwilliamson** [comment]
Okay, so installing `libopenblas-dev` on Ubuntu fixed the issue I was having. The update alternatives command didn't work for me. I wonder why I didn't need this for the previous version?

Anyway, all good here:
```
(v0.7) pkg> build Pardiso
  Building Pardiso → `~/.julia/packages/Pardiso/Sakgl/deps/build.log`

(v0.7) pkg> test Pardiso
   Testing Pardiso
    Status `/tmp/tmpo4zmfZ/Manifest.toml`
  [46dd5b70] Pardiso v0.3.2+ #kc/upgrade (https://github.com/JuliaSparse/Pardiso.jl.git)
  [2a0f44e3] Base64  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Base64`]
  [8ba89e20] Distributed  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Distributed`]
  [b77e0a4c] InteractiveUtils  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/InteractiveUtils`]
  [8f399da3] Libdl  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Libdl`]
  [37e2e46d] LinearAlgebra  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/LinearAlgebra`]
  [56ddb016] Logging  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Logging`]
  [d6f4376e] Markdown  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Markdown`]
  [9a3f8284] Random  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Random`]
  [9e88b42a] Serialization  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Serialization`]
  [6462fe0b] Sockets  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Sockets`]
  [2f01184e] SparseArrays  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/SparseArrays`]
  [8dfed614] Test  [`/opt/julia-0.7.0/bin/../share/julia/stdlib/v0.7/Test`]
Testing DataType[PardisoSolver]
Test Summary: | Pass  Total
solving       |   48     48
[PARDISO]: Matrix type      : general real
[PARDISO]: Matrix dimension : 10
[PARDISO]: Matrix non-zeros : 100
[PARDISO]: Abs. coeff. range: min 1.09e-02  max 9.65e-01
[PARDISO]: RHS no. 1:  min 1.93e-01  max 9.52e-01
[PARDISO]: RHS no. 2:  min 1.34e-02  max 9.83e-01
Test Summary: | Pass  Total
error checks  |    5      5
Test Summary:       | Pass  Total
getters and setters |   11     11
   Testing Pardiso tests passed
```

**Comment by KristofferC** [comment]
Probably https://github.com/JuliaLang/julia/pull/22828

**Comment by ianwilliamson** [comment]
As my above comment shows, the tests all pass but I'm getting this error when trying to solve from my own code:
```
ERROR: MethodError: no method matching unsafe_convert(::Type{Ptr{Int32}}, ::Base.RefValue{Int64})
Closest candidates are:
  unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{T}) where T at refvalue.jl:30
  unsafe_convert(::Type{T<:Ptr}, ::T<:Ptr) where T<:Ptr at essentials.jl:352
  unsafe_convert(::Type{P<:Ptr}, ::Ptr) where P<:Ptr at essentials.jl:353
  ...
Stacktrace:
 [1] pardiso(::Pardiso.PardisoSolver) at /home/-----/.julia/packages/Pardiso/Sakgl/src/project_pardiso.jl:100
```
My code is:
```
        ps = PardisoSolver()
        set_matrixtype!(ps, Pardiso.COMPLEX_NONSYM)
        x = Pardiso.solve(ps, A, b);
        set_phase!(ps, Pardiso.RELEASE_ALL)
        pardiso(ps)
```
Let me know if this is somehow an issue on my end and I can investigate.

**Comment by KristofferC** [comment]
Should be fixed.

**Comment by ianwilliamson** [comment]
Great! That eliminated the error on my end.

I think it would be nice to eventually have a different way of loading the Pardiso binary. After updating to bdf181a with your fix, a new copy of the package was created under a new folder with a randomly generated name (but without the binary under /deps/). Also running the build script after updating didn't provide any indicator of not loading the binary. Maybe there's some neat feature in Pkg3 that would address this?

**Comment by KristofferC** [comment]
Using an `ENV` variable to the path where the library is stored seems easiest for now.

**Comment by KristofferC** [comment]
I updated the install instructions, please see if they are understandable and that they work.

**Comment by ianwilliamson** [comment]
Build seems to work with the `ENV` aproach. I did have to remove the `.so` file extension in order for the build script to pick it up.

**Comment by KristofferC** [comment]
Oh, that is because the Linux names are missing the `.so` extension in the build script.

**Comment by KristofferC** [comment]
Ok, I think this is ready to go now.

In the last commit I added finalizers to the `PardisoType` objects so that when they get garbage collected, they free the memory inside the pardiso library. Previously, I think we just leaked here if the user forgot to manually free it.

Unless I hear of some problems with the latest state of this branch I aim to merge and tag in a day or so.

**Pull request merged**